### PR TITLE
Rds to s3 script

### PR DESCRIPTION
**Pull request by Shurc1val**
Additions:
- `rds_to_s3.py` python file getting any records older than 24hrs from recording/watering tables, adding them to relevant s3 files, and removing from the db; dry tested against empty database and s3 (it runs), but **not tested with data**; **not yet documented**.
- `pipeline/` folder to contain `rds_to_s3.py`
- `setup_local_db_connection.sh` shell file containing instructions and code to allow local connection to database to be set up
- `connect_to_db.sh`shell file to connect to db

Changes:
- `load.py` and `extract.py` moved into `pipeline/` folder

### CODE REVIEW

File: setup_local_db_connection.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Must be run before this file:
```
Comment (caitlinturnidge) [inline setup_local_db_connection.sh:1]: Nice one! I would have forgotten to add this file

File: pipeline/rds_to_s3.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    """
+    Gets any records from watering and recording tables in df older than 24hours, combines with
+    the csv files in s3 bucket for current day (if any exist), and saves the result as csvs in the
+    bucket with a name ending in _{day}.csv.
+    """
+    db_engine = get_database_engine()
+    db_connection = db_engine.connect()
+    db_metadata = db.MetaData(schema=environ['DB_SCHEMA'])
+
+    s3_client = client("s3",
```
Comment (caitlinturnidge) [inline pipeline/rds_to_s3.py:112]: could put this in a get_s3_client() function here?

File: pipeline/rds_to_s3.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                       aws_access_key_id=environ['AWS_ACCESS_KEY_ID'],
+                       aws_secret_access_key=environ['AWS_SECRET_ACCESS_KEY'])
+
+    for data_type in ['recording', 'watering']:
+        df = get_old_records(data_type, db_engine, db_connection, db_metadata)
+        df = pd.concat([df, get_current_csv_data(data_type, s3_client)])
+        upload_to_s3(data_type, df, s3_client)
+        delete_oldest_records(data_type, db_engine, db_connection, db_metadata)
+
```
Comment (caitlinturnidge) [inline pipeline/rds_to_s3.py:121]: good practice to add a db_connection.close() at the end

**Review by caitlinturnidge — DISMISSED** [review]

File: pipeline/rds_to_s3.py (hunk 3 above)
Comment (Shurc1val) [reply pipeline/rds_to_s3.py:121]: Good point; adding now

File: pipeline/rds_to_s3.py (hunk 2 above)
Comment (Shurc1val) [reply pipeline/rds_to_s3.py:112]: Personally, because it's a single line and is only used once, I wouldn't

File: connect_to_db.sh [hunk 4]
```diff
@@ -0,0 +1,3 @@
+source .env
```
Comment (david-nepo) [inline connect_to_db.sh:1]: 🐚

File: pipeline/rds_to_s3.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            f"mssql+pymssql://{environ['DB_USER']}:{environ['DB_PASSWORD']}@{environ['DB_HOST']}:{environ['DB_PORT']}/{environ['DB_NAME']}?charset=utf8"
+        )
+        return engine
+    except Exception as e:
+        print(f"Error creating database engine: {e}")
+        raise e
+
+
+def get_old_records(table_name: str, db_engine: db.Engine, connection: Connection, metadata,
+                    date_cutoff: str = STR_DAY_AGO):
```
Comment (david-nepo) [inline pipeline/rds_to_s3.py:34]: Nitpick - but PEP8 says to not put spaces around '=' in arguments...

[Don’t use spaces around the = sign when used to indicate a keyword argument, or when used to indicate a default value for an unannotated function parameter:](https://peps.python.org/pep-0008/#other-recommendations)

```
# Correct:
def complex(real, imag=0.0):
    return magic(r=real, i=imag)
```

```
# Wrong:
def complex(real, imag = 0.0):
    return magic(r = real, i = imag)
```

File: connect_to_db.sh (hunk 4 above)
Comment (Shurc1val) [reply connect_to_db.sh:1]: ...what? 😅

File: pipeline/rds_to_s3.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        query = db.select(table).where(table.columns.datetime < str.encode(date_cutoff))
+        response = connection.execute(query)
+        return pd.DataFrame(response.fetchall())
+
+    except Exception as e:
+        raise e
+
+
+def get_day_bucket_keys(s3_client: client, folder_path: str, day: int = TODAY.day,
+                    bucket_name: str = environ['BUCKET_NAME']) -> list:
```
Comment (david-nepo) [inline pipeline/rds_to_s3.py:50]: Same spacing thing here

File: connect_to_db.sh (hunk 4 above)
Comment (Shurc1val) [reply connect_to_db.sh:1]: Ohhh, shell, I get it

File: pipeline/rds_to_s3.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    if objects:
+        return [obj['Key'] for obj in objects if obj['Key'].split('_')[-1] == f'{day}.csv']
+    return []
+
+
+def get_current_csv_data(data_type: str, s3_client: client, bucket_name:
+                         str = environ['BUCKET_NAME'], date: str = TODAY) -> pd.DataFrame:
+    """
+    Downloads relevant files of specified data_type (watering/recording) from S3 to local, and
+    then returns it as a pandas dataframe
```
Comment (david-nepo) [inline pipeline/rds_to_s3.py]: Needs a full stop at end of doc string to be consistent with others

File: pipeline/rds_to_s3.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        raise e
+
+
+def get_old_records(table_name: str, db_engine: db.Engine, connection: Connection, metadata,
+                    date_cutoff: str = STR_DAY_AGO):
+    """
+    Retrieves records older than 24 hours (by attribute 'datetime') from db table with given name
+    (watering/recording).
+    """
+    try:
```
Comment (david-nepo) [inline pipeline/rds_to_s3.py:39]: Nice functionality

File: pipeline/rds_to_s3.py (hunk 5 above)
Comment (Shurc1val) [reply pipeline/rds_to_s3.py:34]: That's only the case if there's no type hints; from directly below the bit you've attached:

Screenshot 2023-12-19 at 13 55 11

**Review by david-nepo — DISMISSED** [review]
All looks good, apart from nitpick on keyword list spacing. Happy to approve and merge now and then clean up later on prior to submission. Nice!

File: pipeline/rds_to_s3.py (hunk 6 above)
Comment (Shurc1val) [reply pipeline/rds_to_s3.py:50]: Same response 🙃

File: pipeline/rds_to_s3.py (hunk 7 above)
Comment (Shurc1val) [reply pipeline/rds_to_s3.py]: Very true, good spot

**Review by david-nepo — DISMISSED** [review]
big fat approve

**Review by david-nepo — APPROVED** [review]
You shall ~not~ pass
